### ISSUE TITLE
Have to publish with subscription name in the object

### ISSUE THREAD

**Issue by AnthonyCC** [apollographql/graphql-subscriptions#51]
I was following the tutorail in the README file. I noticed if I use 
```
// publish to the channel
pubsub.publish('newCommentsChannel', {
  id: 123,
  content: 'Test',
  repoFullName: 'apollostack/GitHunt-API',
  posted_by: 'helfer',
});
```

I will get
```
{
  data: {
     commentAdded: null
   }
}

But if I publish with the subscription name in the object

// publish to the channel
pubsub.publish('newCommentsChannel', {
  commentAdded: {
    id: 123,
    content: 'Test',
    repoFullName: 'apollostack/GitHunt-API',
    posted_by: 'helfer',
  }
});
```

Then I will get the result as in the README.
Am I missing something from the document?

**Comment by NeoPhi** [comment]
Can you post a link to a repo that duplicates this issue? Depending on how you are constructing your schema you may have to attach a custom resolver to the `subscription.commentAdded` field.

**Comment by DxCx** [comment]
the implementation is abit bugy,
each time a new value emits, it will run graphql query and return the result.
when running the graphql query, it will pass the emitted value from pubsub as root.
the right thing to do if you don't want to write a resolver is make sure the root will be passed as:
{ channelName: value } and not just value.

**Comment by AnthonyCC** [comment]
Sure, this is the link to my repo. https://github.com/AnthonyCC/graphQL-Demo

The schema is defined in /server/graphQL/schema.js.

The data is published to the channel in /server/route/graphQL.js, `updateUserPassword`.

Thanks for looking into this issue.

**Comment by NeoPhi** [comment]
In your example the object you are publishing is nested incorrectly.
`sub.pubsub.publish('passwordUpdatedChannel', {passwordUpdated:user});`
should be:
`sub.pubsub.publish('passwordUpdatedChannel', user);`

This example shows the behavior:
```
const graphql = require('graphql');
const sub = require('graphql-subscriptions');

const schema = graphql.buildSchema(`
  type Subscription {
    passwordUpdated: User
  }
  type Query {
    getMessage: String,
  }
  type User {
    name: String,
    password: String
  }
`);

const pubsub = new sub.PubSub();
const manager = new sub.SubscriptionManager({
  schema,
  pubsub,
  setupFunctions: {
    passwordUpdated: (options, args) => ({
      passwordUpdatedChannel: { filter: user => user != null },
    }),
  },
});

manager.subscribe({
  query: `subscription passwordUpdated { passwordUpdated { name password } }`,
  context: {},
  callback: (err, data) => console.log('callback', err, data),
});

pubsub.publish('passwordUpdatedChannel', {
  passwordUpdated: { name: 'name', password: 'password' },
});
```

**Comment by n1ru4l** [comment]
I had the same issue and fixed it by adding a Subscription resolver to my resolver map

In you case this woule be
```javascript
const resolvers = {
  Subscription: {
    commentAdded: comment => comment,
  },
}
```

In the [documentation](http://dev.apollodata.com/tools/graphql-subscriptions/subscriptions-to-schema.html) it says:

> Create a resolver just like queries and mutations if you need to perform a logic over the published data

For me it did not work at all without the resolver. If someone can confirm this I am going submit a PR.

**Comment by AnthonyCC** [comment]
Yes. @NeoPhi , that is the issue I am having. If I publish the data with `sub.pubsub.publish('passwordUpdatedChannel', user);`

Then I won't get the data in the subscriber, I will get 
```
{
  data: {
     passwordUpdatedChannel: null
   }
}
```

@n1ru4l , adding a subscription resolver does resolve the issue for me. Thanks for sharing your finding.

I have two questions
1) Do I need to have the subscription resolver even though I don't need perform any logic over the published data, other than just return the data?
2) Is there a way to add subscription resolver in `graphql-js` (which is the library I am using on the server side)? Or I have to use the Apollo's one?

**Comment by NeoPhi** [comment]
@AnthonyCC When I run the code listed in #issuecomment-291157614 the output is:
`callback null { data: { passwordUpdated: { name: 'name', password: 'password' } } }`
So I'm unclear what you mean by the subscriber not getting the data.

**Comment by AnthonyCC** [comment]
@NeoPhi ,

If you publish the object with the subscription name nested in the object, then you will get the expected  output, which is`callback null { data: { passwordUpdated: { name: 'name', password: 'password' } } }` . As in you example listed, 
```
pubsub.publish('passwordUpdatedChannel', {
  passwordUpdated: { name: 'name', password: 'password' },
});
```
You have to nest the object with `passwordUpdated`

However, if you publish the data by `sub.pubsub.publish('passwordUpdatedChannel', user);`, then you will get `callback null { data: { passwordUpdated: null } }`

**Comment by julioolvr** [comment]
I found this while struggling with something similar - I wanted to use a certain property in the `filter` option that I didn't want to publish over the subscription. In my case the behavior described here was actually useful: I was able to `publish` an object with the real subscription data under the subscription name, and some extra properties that didn't get sent through the subscription.

That being said, it's too implicit imo. Maybe it'd be nice to have a `map` option (just like we have `filter`) where we can convert whatever we publish to pubsub into something else to send as the subscription result.

**Comment by Ancinek** [comment]
Finally. Found this issue and the problem that I was trying to solve for more than an hour is gone. None of the tutorials mentioned that you have to nest it inside the `publish` method and match the key with the subscription name.

**Comment by mjasnikovs** [comment]
Please add this to **readme**! I spent hours...

**Comment by grantwwu** [comment]
Keeping this open because it seems like we could use more documentation on this.
